pallets/xyk/src/lib.rs

```diff
@@ -948,6 +875,220 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, Balance> for Pallet<T> {
+
+		let amount_out_with_fee = amount_out
+			.amount
+			.checked_add(amount_out.fee)
+			.ok_or(ExecutorError::Error(()))?;
+
+		let amount_in =
+			hydra_dx_math::xyk::calculate_in_given_out(asset_out_reserve, asset_in_reserve, amount_out_with_fee)
+				.map_err(|_| ExecutorError::Error(()))?;
+
+		let transfer_fee = Self::calculate_fee(amount_in).map_err(|_| ExecutorError::Error(()))?;
+
+		Ok(TradeCalculation {
+			amount: amount_in,
+			fee: transfer_fee,
+		})
+	}
+
+	fn execute_sell(
+		pool_type: PoolType<AssetId>,
+		who: &T::AccountId,
+		asset_in: AssetId,
+		asset_out: AssetId,
+		amount_in: Self::TradeCalculationResult,
+	) -> Result<(), ExecutorError<Self::Error>> {
+		if pool_type != PoolType::XYK {
+			return Err(ExecutorError::NotSupported);
+		}
+
+		let amount_out = Self::calculate_sell(pool_type, asset_in, asset_out, amount_in)?;
+
+		let assets = AssetPair { asset_in, asset_out };
+		let fee = (asset_out, amount_out.fee);
+		let amount_in_without_fee = amount_in
+			.amount
+			.checked_sub(amount_in.fee)
+			.ok_or(ExecutorError::Error(()))?;
+		let amount_out_without_fee = amount_out
+			.amount
+			.checked_sub(amount_out.fee)
+			.ok_or(ExecutorError::Error(()))?;
+
+		let amm_transfer: AMMTransfer<T::AccountId, AssetId, AssetPair, Balance> = AMMTransfer {
+			origin: who.clone(),
+			assets,
+			amount: amount_in_without_fee,
+			amount_out: amount_out_without_fee,
+			fee,
+			discount: false,
+			discount_amount: 0,
+		};
+		//TODO: Dani - proper error propagation
+		Self::do_validate_sell_common(amount_in_without_fee, assets, &who).map_err(|_| ExecutorError::Error(()))?;
+		Self::do_execute_sell(&amm_transfer).map_err(|_| ExecutorError::Error(()))?;
+
+		Ok(())
+	}
+
+	fn execute_buy(
+		pool_type: PoolType<AssetId>,
+		who: &T::AccountId,
+		asset_in: AssetId,
+		asset_out: AssetId,
+		amount_out: Self::TradeCalculationResult,
+	) -> Result<(), ExecutorError<Self::Error>> {
+		if pool_type != PoolType::XYK {
+			return Err(ExecutorError::NotSupported);
+		}
+
+		let amount_in = Self::calculate_buy(pool_type, asset_in, asset_out, amount_out)?;
+
+		let assets = AssetPair { asset_in, asset_out };
+		let fee = (asset_in, amount_in.fee);
+		let amount_out_with_fee = amount_out
+			.amount
+			.checked_add(amount_out.fee)
+			.ok_or(ExecutorError::Error(()))?;
+
+		let amm_transfer: AMMTransfer<T::AccountId, AssetId, AssetPair, Balance> = AMMTransfer {
+			origin: who.clone(),
+			assets,
+			amount: amount_out_with_fee,
+			amount_out: amount_in.amount,
+			fee,
+			discount: false,
+			discount_amount: 0,
+		};
+		Self::do_validate_buy_common(assets, amount_out_with_fee).map_err(|_| ExecutorError::Error(()))?;
+		Self::do_execute_buy(&amm_transfer).map_err(|_| ExecutorError::Error(()))?;
+
+		Ok(())
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn do_validate_sell_common(amount: Balance, assets: AssetPair, who: &T::AccountId) -> Result<(), DispatchError> {
```

Comment:
what is the meaning of the `_common` suffix?

Reply:
I just wanted to clearly separate it from the main function `validate_sell`, so it is a common logic used in multiple places. But you are right, having a suffix can cause confusion for the code readers. Probably just `do_validate_sell` alone does the job.